src/thread/thread_worker_pool_taskcap.hpp

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+//
+// cap the number of tasks that are pushed into an worker pool and block incoming pushes when maxed
+//
+
+#ifndef _THREAD_WORKER_POOL_TASKCAP_HPP_
+#define _THREAD_WORKER_POOL_TASKCAP_HPP_
+
+#include "thread_worker_pool.hpp"
+
+namespace cubthread
+{
+  template <typename Context>
+  class worker_pool_task_capper
+  {
+    private:
+      using context_type = Context;
+      using task_type = task<Context>;
+      using worker_pool_type = worker_pool<Context>;
+
+    public:
+      worker_pool_task_capper (worker_pool<Context> *worker_pool);
+      ~worker_pool_task_capper ();
+
+      void push_task (task<Context> *task);
+      cubthread::worker_pool<Context> *get_worker_pool ();
+      void end_task ();
```

Comment:
I think better will be to limit end_task function to private access, in order to avoid being called